workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/WorkflowRendering.kt

```diff
@@ -0,0 +1,201 @@
+ * from Composable functions.
+ *
+ * ## Example
+ *
+ * ```
+ * data class FramedRendering(
+ *   val borderColor: Color,
+ *   val child: Any
+ * )
+ *
+ * val FramedContainerViewFactory = composeViewFactory<FramedRendering> { rendering, environment ->
+ *   Surface(border = Border(rendering.borderColor, 8.dp)) {
+ *     WorkflowRendering(rendering.child, environment)
+ *   }
+ * }
+ * ```
+ *
+ * @param rendering The workflow rendering to display. May be of any type for which a [ViewFactory]
+ * has been registered in this environment's [ViewRegistry].
+ * @param modifier A [Modifier] that will be applied to composable used to show [rendering].
+ *
+ * @throws IllegalArgumentException if no factory can be found for [rendering]'s type.
+ */
+@WorkflowUiExperimentalApi
+@Composable public fun WorkflowRendering(
+  rendering: Any,
+  viewEnvironment: ViewEnvironment,
+  modifier: Modifier = Modifier
+) {
+  val viewRegistry = remember(viewEnvironment) { viewEnvironment[ViewRegistry] }
+
+  // This will fetch a new view factory any time the new rendering is incompatible with the previous
+  // one, as determined by Compatible. This corresponds to WorkflowViewStub's canShowRendering
+  // check.
+  // We also use the ViewRegistry as a key, in case a new one gets passed down and we need to ask
+  // it for a more current factory.
+  // The view registry may return a new factory instance for a rendering every time we ask it, for
+  // example if an AndroidViewRendering doesn't share its factory between rendering instances. We
+  // intentionally don't ask it for a new instance every time to match the behavior of
+  // WorkflowViewStub and other containers, which only ask for a new factory when the rendering is
+  // incompatible.
+  val renderingKey = Compatible.keyFor(rendering)
+  val viewFactory = remember(renderingKey, viewRegistry) {
+    viewRegistry.getFactoryForRendering(rendering)
+  }
+
+  // By surrounding the rendering composable with this key function, any time the new rendering is
+  // not compatible with the previous rendering we'll tear down the previous subtree of the
+  // composition, including its lifecycle, which destroys the lifecycle and any remembered state.
+  // If the view factory created an Android view, this will also remove the old one from the view
+  // hierarchy before replacing it with the new one.
+  key(viewFactory) {
+    // Just like WorkflowViewStub, we need to manage a Lifecycle for the child view. We just provide
+    // a local here – ViewFactoryAndroidView will handle setting the appropriate view tree owners
+    // on the child view when necessary. Because this function is surrounded by a key() call, when
+    // the rendering is incompatible, the lifecycle for the old view will be destroyed.
+    val lifecycleOwner = rememberChildLifecycleOwner()
+    CompositionLocalProvider(LocalLifecycleOwner provides lifecycleOwner) {
+      if (viewFactory is ComposeViewFactory) {
```

Comment:
Instead of having this if/else here, we could make a function `ViewFactory.asComposeViewFactory` that returns this or creates a wrapper factory that has the Android view integration logic, then this function wouldn't have to special case anything. I think that might be a bit cleaner? Or is it just over-engineered?

Reply:
I think I've convinced myself this is a good idea, since it makes extra, crystal clear the separation between what is general logic for managing the child, and what is specialized logic for handling Android view integration.

Reply:
Sorry: convinced yourself having the case in `WorkflowRendering` here is good? or having the extension on the `ViewFactory` is good?

Reply:
Sorry, pre-coffee comment. I've convinced myself I like the `asComposeFactory`. And just to be extra clear:

<table>
<tr><th>
With `asComposeViewFactory()`
</th><th>
With `if/else`
</th></tr>
<tr>
<td>

```kotlin
val factory = remember {
  registry.getFactoryForRendering(rendering)
   .asComposeViewFactory()
}

key(…) {
  CompositionLocalProvider(…) {
    Box(…) {
      factory.Content(rendering, viewEnvironment)
    }
  }
}
```

</td><td>

```kotlin
val factory = remember {
  registry.getFactoryForRendering(rendering)
}

key(…) {
  CompositionLocalProvider(…) {
    if (factory is ComposeViewFactory) {
      Box(…) {
        factory.Content(rendering, viewEnvironment)
      }
    } else {
      AndroidViewThingy(…)
    }
  }
}
```

</td>
</tr>
</table>

And `asComposeViewFactory()` would look like this:
```kotlin
private fun <R : Any> ViewFactory<R>.asComposeViewFactory() = 
  (this as? ComposeViewFactory) ?: object : ComposeViewFactory<R> {
    @Composable override fun Content(
      rendering: R,
      viewEnvironment: ViewEnvironment
    ) {
      AndroidViewThingy(rendering, viewEnvironment)
    }
  }
```

Reply:
+1 I like `.asComposeViewFactory()` but I think it deserves some of your good documentation so folks know what is going on.

Reply:
Oh hell yeah. I really like `.asComposeViewFactory()`, sold.

Reply:
I'll break out the top-shelf docs for this one